File: backend/app/services/like_service.py
"""点赞服务：动态与评论的点赞切换。"""

from __future__ import annotations

from sqlalchemy import func
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.models.comments import Comment
from app.models.likes import Like
from app.models.posts import Post
from app.models.users import User
from app.schemas.comment import LikeCountOut
from app.schemas.common import AppError, ErrorCode
from app.utils.visibility import can_view_comment, can_view_post


def toggle_post_like(db: Session, user: User, post_id: int) -> dict:
    post = db.query(Post).filter(Post.id == post_id, Post.deleted_at.is_(None)).first()
    if not post:
        raise AppError(ErrorCode.NOT_FOUND, "Post not found", 404)
    if not can_view_post(db, user.id, post):
        raise AppError(ErrorCode.FORBIDDEN, "No permission to like this post", 403)

    _toggle(db, user, "post", post_id)

    like_count = (
        db.query(func.count(Like.id))
        .filter(Like.target_type == "post", Like.target_id == post_id)
        .scalar()
    ) or 0
    liked_by_me = (
        db.query(Like.id)
        .filter(
            Like.target_type == "post",
            Like.target_id == post_id,
            Like.user_id == user.id,
        )
        .first()
    ) is not None

    return LikeCountOut(
        target_type="post",
        target_id=post_id,
        like_count=like_count,
        liked_by_me=liked_by_me,
    ).model_dump(mode="json")


def toggle_comment_like(db: Session, user: User, comment_id: int) -> dict:
    comment = (
        db.query(Comment).filter(Comment.id == comment_id, Comment.deleted_at.is_(None)).first()
    )
    if not comment:
        raise AppError(ErrorCode.COMMENT_NOT_FOUND, "Comment not found", 404)

    post = db.query(Post).filter(Post.id == comment.post_id, Post.deleted_at.is_(None)).first()
    if not post:
        raise AppError(ErrorCode.NOT_FOUND, "Post not found", 404)
    if not can_view_post(db, user.id, post):
        raise AppError(ErrorCode.FORBIDDEN, "No permission to like this comment", 403)
    if not can_view_comment(db, user.id, comment, post):
        raise AppError(ErrorCode.FORBIDDEN, "No permission to like this comment", 403)

    _toggle(db, user, "comment", comment_id)

    like_count = (
        db.query(func.count(Like.id))
        .filter(Like.target_type == "comment", Like.target_id == comment_id)
        .scalar()
    ) or 0
    liked_by_me = (
        db.query(Like.id)
        .filter(
            Like.target_type == "comment",
            Like.target_id == comment_id,
            Like.user_id == user.id,
        )
        .first()
    ) is not None

    return LikeCountOut(
        target_type="comment",
        target_id=comment_id,
        like_count=like_count,
        liked_by_me=liked_by_me,
    ).model_dump(mode="json")


def _toggle(db: Session, user: User, target_type: str, target_id: int) -> None:
    existing = (
        db.query(Like)
        .filter(
            Like.target_type == target_type,
            Like.target_id == target_id,
            Like.user_id == user.id,
        )
        .first()
    )

    if existing:
        db.delete(existing)
        db.commit()
    else:
        like = Like(
            target_type=target_type,
            target_id=target_id,
            user_id=user.id,
        )
        db.add(like)
        try:
            db.commit()
        except IntegrityError:
            db.rollback()
            existing = (
                db.query(Like)
                .filter(
                    Like.target_type == target_type,
                    Like.target_id == target_id,
                    Like.user_id == user.id,
                )
                .first()
            )
            if existing:
                db.delete(existing)
                db.commit()

File: backend/internal/service/like_service.go
package service

import (
	"context"

	"backend/internal/model"
	"backend/internal/repository"
)

type LikeService struct {
	repos *repository.Repositories
}

func NewLikeService(repos *repository.Repositories) *LikeService {
	return &LikeService{repos: repos}
}

func (s *LikeService) TogglePostLike(ctx context.Context, user *model.User, postID int) (*model.LikeCountOut, error) {
	post, err := s.repos.Post.GetByID(ctx, postID)
	if err != nil {
		return nil, err
	}
	if post == nil {
		return nil, model.NewAppError(model.ErrNotFound, "动态不存在", 404)
	}

	canView, err := s.canViewPost(ctx, user.ID, post)
	if err != nil {
		return nil, err
	}
	if !canView {
		return nil, model.NewAppError(model.ErrForbidden, "无权点赞此动态", 403)
	}

	liked, count, err := s.repos.Like.Toggle(ctx, "post", postID, user.ID)
	if err != nil {
		return nil, err
	}

	return &model.LikeCountOut{
		TargetType: "post",
		TargetID:   postID,
		LikeCount:  count,
		LikedByMe:  liked,
	}, nil
}

func (s *LikeService) UnlikePost(ctx context.Context, user *model.User, postID int) (*model.LikeCountOut, error) {
	post, err := s.repos.Post.GetByID(ctx, postID)
	if err != nil {
		return nil, err
	}
	if post == nil {
		return nil, model.NewAppError(model.ErrNotFound, "动态不存在", 404)
	}

	canView, err := s.canViewPost(ctx, user.ID, post)
	if err != nil {
		return nil, err
	}
	if !canView {
		return nil, model.NewAppError(model.ErrForbidden, "无权操作", 403)
	}

	count, err := s.repos.Like.Unlike(ctx, "post", postID, user.ID)
	if err != nil {
		return nil, err
	}

	return &model.LikeCountOut{
		TargetType: "post",
		TargetID:   postID,
		LikeCount:  count,
		LikedByMe:  false,
	}, nil
}

func (s *LikeService) ToggleCommentLike(ctx context.Context, user *model.User, commentID int) (*model.LikeCountOut, error) {
	comment, err := s.repos.Comment.GetByID(ctx, commentID)
	if err != nil {
		return nil, err
	}
	if comment == nil {
		return nil, model.NewAppError(model.ErrCommentNotFound, "评论不存在", 404)
	}

	post, err := s.repos.Post.GetByID(ctx, comment.PostID)
	if err != nil {
		return nil, err
	}
	if post == nil {
		return nil, model.NewAppError(model.ErrNotFound, "动态不存在", 404)
	}

	canViewP, err := s.canViewPost(ctx, user.ID, post)
	if err != nil {
		return nil, err
	}
	if !canViewP {
		return nil, model.NewAppError(model.ErrForbidden, "无权点赞此评论", 403)
	}

	canViewC, err := s.canViewComment(ctx, user.ID, comment, post)
	if err != nil {
		return nil, err
	}
	if !canViewC {
		return nil, model.NewAppError(model.ErrForbidden, "无权点赞此评论", 403)
	}

	liked, count, err := s.repos.Like.Toggle(ctx, "comment", commentID, user.ID)
	if err != nil {
		return nil, err
	}

	return &model.LikeCountOut{
		TargetType: "comment",
		TargetID:   commentID,
		LikeCount:  count,
		LikedByMe:  liked,
	}, nil
}

func (s *LikeService) UnlikeComment(ctx context.Context, user *model.User, commentID int) (*model.LikeCountOut, error) {
	comment, err := s.repos.Comment.GetByID(ctx, commentID)
	if err != nil {
		return nil, err
	}
	if comment == nil {
		return nil, model.NewAppError(model.ErrCommentNotFound, "评论不存在", 404)
	}

	post, err := s.repos.Post.GetByID(ctx, comment.PostID)
	if err != nil {
		return nil, err
	}
	if post == nil {
		return nil, model.NewAppError(model.ErrNotFound, "动态不存在", 404)
	}

	canViewP, err := s.canViewPost(ctx, user.ID, post)
	if err != nil {
		return nil, err
	}
	if !canViewP {
		return nil, model.NewAppError(model.ErrForbidden, "无权操作", 403)
	}

	count, err := s.repos.Like.Unlike(ctx, "comment", commentID, user.ID)
	if err != nil {
		return nil, err
	}

	return &model.LikeCountOut{
		TargetType: "comment",
		TargetID:   commentID,
		LikeCount:  count,
		LikedByMe:  false,
	}, nil
}

func (s *LikeService) canViewPost(ctx context.Context, viewerID int, post *model.Post) (bool, error) {
	if post.DeletedAt != nil {
		return false, nil
	}
	if post.Visibility == "public" {
		return true, nil
	}
	if viewerID == post.UserID {
		return true, nil
	}
	return s.repos.Friend.AreFriends(ctx, viewerID, post.UserID)
}

func (s *LikeService) canViewComment(ctx context.Context, viewerID int, comment *model.Comment, post *model.Post) (bool, error) {
	if comment.DeletedAt != nil || post.DeletedAt != nil {
		return false, nil
	}
	if post.Visibility == "public" {
		return true, nil
	}
	if comment.UserID == viewerID || comment.UserID == post.UserID {
		return true, nil
	}
	return s.repos.Friend.AreFriends(ctx, viewerID, comment.UserID)
}
